modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/GridTcpCommunicationInverseConnectionEstablishingTest.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.EnvironmentType;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.util.nio.GridCommunicationClient;
+import org.apache.ignite.lang.IgnitePredicate;
+import org.apache.ignite.spi.discovery.tcp.internal.TcpDiscoveryNode;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.cache.CacheAtomicityMode.ATOMIC;
+import static org.apache.ignite.cache.CacheAtomicityMode.TRANSACTIONAL;
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.FULL_SYNC;
+
+/**
+ *
+ */
+public class GridTcpCommunicationInverseConnectionEstablishingTest extends GridCommonAbstractTest {
+    /** */
+    private static final IgnitePredicate<ClusterNode> CLIENT_PRED = c -> c.isClient();
+
+    /** */
+    private static final String UNREACHABLE_IP = "172.31.30.132";
+
+    /** */
+    private static final String UNRESOLVED_HOST = "unresolvedHost";
+
+    /** */
+    private boolean clientMode;
+
+    /** */
+    private EnvironmentType envType = EnvironmentType.STAND_ALONE;
+
+    /** */
+    private CacheConfiguration[] ccfgs;
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setFailureDetectionTimeout(8_000);
+
+        cfg.setActiveOnStart(true);
+
+        cfg.setCommunicationSpi(new TestCommunicationSpi());
+
+        if (ccfgs != null) {
+            cfg.setCacheConfiguration(ccfgs);
+
+            ccfgs = null;
+        }
+
+        if (clientMode) {
+            cfg.setEnvironmentType(envType);
+            cfg.setClientMode(true);
+        }
+
+        return cfg;
+    }
+
+    @Test
+    public void testUnreachableClientInVirtualizedEnvironment() throws Exception {
+        executeCacheTestWithUnreachableClient(EnvironmentType.VIRTUALIZED);
+    }
+
+    @Test
+    public void testUnreachableClientInStandAloneEnvironment() throws Exception {
+        executeCacheTestWithUnreachableClient(EnvironmentType.STAND_ALONE);
+    }
+
+    /**
+     *
+     * @param envType
+     * @throws Exception
+     */
+    private void executeCacheTestWithUnreachableClient(EnvironmentType envType) throws Exception {
+        int srvs = 3;
+
+        for (int i = 0; i < srvs; i++) {
+            ccfgs = cacheConfigurations1();
+
+            startGrid(i);
+        }
+
+        clientMode = true;
+        this.envType = envType;
+
+        startGrid(srvs);
```

Comment:
You can call "startClientGrid" and check configuration instead of field. So the field could be removed.